### PR TITLE
Update payment request when billing data changes

### LINKED ISSUE #1830: Express Payment Buttons are broken after updating products in the Cart Block
**Describe the bug**

Testing against the latest version of Stripe (5.4.1) and Blocks (going back as far as v5.0), there is a new incompatibility with the Cart Block and express payments. When the cart changes, the payment request object also updates, but the payment modal does not show updated totals. Additionally, after a cart change, the "cancel" handler fails to run leaving the payment button in a loading state.

I tested this on Edge (chromium) and Chrome. Both showed the same problem. I am marking this as high priority because functionality is broken.

**To Reproduce**
1. Go to a page showing the Cart Block, with items in your cart
2. Click the express payment button
3. Notice the modal shows the correct total. Cancel payment. Notice the button goes back to a clickable state.
4. Change the qty of an item in the cart and wait a few moments for the change to persist.
5. Click the express payment button
6. **Notice the total is wrong.**
7. Click the cancel button in the payment request modal. **Notice the button remains in loading state.**

**Expected behavior**
After updating cart items, the express payment button needs to reflect the new totals.

**Environment (please complete the following information):**
 - WordPress 5.8
 - WooCommerce 5.7.0-dev
 - Stripe 5.4.1
 - Edge/Chrome Browser
 - WooCommerce Blocks 5.9.0-dev

### PR DESCRIPTION
**Pull request by reykjalin**
Fixes #1830 

# Changes proposed in this Pull Request:

- Update the `PaymentRequest` object when billing data changes in the cart block instead of recreating it.

The `PaymentRequestButtonElement` doesn't support re-assigning the `paymentRequest` prop, and as such we have to update the payment request instead.

We achieve this by sending an API request for the cart information to make sure all normalizations are correctly applied, and use that data to update the `PaymentRequest` object.

# Testing instructions

## Make sure Cart Block updates are applied to the Payment Request

> **Note:** This test requires that you have the [WooCommerce Blocks](https://wordpress.org/plugins/woo-gutenberg-products-block/) extension installed on your store.

1. Add a product to your cart.
2. Go to the page containing the Cart Block.
3. Click the Payment Request Button (Apple Pay/Google Pay/Chrome Payment Request).
4. Note the list of items and the total.
5. Click **Cancel** to cancel the payment.
6. Change the number of items in the cart.
7. Click the Payment Request Button again.
8. Make sure the list of items and the total have been updated.

## Make sure Payment Requests can be cancelled multiple times in the Cart Block

> **Note:** This test requires that you have the [WooCommerce Blocks](https://wordpress.org/plugins/woo-gutenberg-products-block/) extension installed on your store.

1. Add a product to your cart.
2. Go to the page containing the Cart Block.
3. Click the Payment Request Button (Apple Pay/Google Pay/Chrome Payment Request).
4. Click **Cancel** to cancel the payment.
5. Click the Payment Request Button again.
6. Click **Cancel** to cancel the payment.
7. Change the number of items in the cart.
8. Click the Payment Request Button again.
9. Click **Cancel** to cancel the payment.
10. Click the Payment Request Button again.
11. Click **Cancel** to cancel the payment.
12. Make sure you can interact with the Payment Request Button (i.e. button is clickable, and not greyed out).
13. Change the number of items in the cart.
14. Again, make sure the Payment Request Button is still active (i.e. button is clickable, and not greyed out).

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

### CODE REVIEW

**Review by diegocurbelo — APPROVED** [review]
The code changes look good and both cases test well.

Just one small detail... if you click the `Buy now` as soon as is re-enabled you get the following js error in the console, and the amount is not updated, If you wait 1 or 2 seconds it works fine; maybe is my test env, and is not a real-life issue just wanted to rise the issue just in case
Screen Shot 2021-10-19 at 18 37 49

**Comment by reykjalin** [comment]
Nice catch @diegocurbelo ! I'm looking into how we can disable the PRB until the cart request has completed, but wasn't able to figure it out on my own. I'll ask Team Rubik in p1634710351348900-slack-C8X6Q7XQU and see if they can help us 👍

**Comment by reykjalin** [comment]
@diegocurbelo I've fixed the issue you found, but the fix is a bit hacky. I couldn't figure out a better way to accomplish this. I'm very open to ideas — there's also more discussion in the Team Rubik thread in Slack 🙂

File: client/blocks/payment-request/payment-request-express.js [hunk 1]
```diff
@@ -120,15 +121,31 @@ const PaymentRequestExpressComponent = ( {
...
+			}
+		>
+			<PaymentRequestButtonElement
+				onClick={ onPaymentRequestButtonClick }
+				options={ {
+					style: paymentRequestButtonStyle,
+					paymentRequest,
+				} }
+			/>
+		</div>
```
Comment (diegocurbelo) [inline client/blocks/payment-request/payment-request-express.js]: How about doing something like this? (very very hacky I know 😞)...

The benefit of this approach is that we keep the button "disabled" until is clickable... I know is just 1 second and maybe not worth it, so feel free to disregard this suggestion

```suggestion
		<div
			{ ...( isUpdatingPaymentRequest && {
				className: 'wc-block-components-loading-mask',
			} ) }
		>
			<div
				{ ...( isUpdatingPaymentRequest && {
					className: 'wc-block-components-loading-mask__children',
				} ) }
			>
				<PaymentRequestButtonElement
					onClick={ onPaymentRequestButtonClick }
					options={ {
						style: paymentRequestButtonStyle,
						paymentRequest,
					} }
				/>
			</div>
		</div>
```

Comment (reykjalin) [reply client/blocks/payment-request/payment-request-express.js]: Works like a charm, although I added the child element class directly onto the Stripe Element itself 🙂 

Fixed in afa9bf6b259ebaa4324d023965a8d7f092af15c1!

Comment (reykjalin) [reply client/blocks/payment-request/payment-request-express.js]: Nice suggestion by the way @diegocurbelo ! I couldn't figure out the necessary class combinations for this 👏

**Review by diegocurbelo — APPROVED** [review]
Works perfectly @reykjalin... good job! 🚢 🚢

**Comment by reykjalin** [comment]
Added a changelog entry, and added the testing instructions to the [5.8.0 release testing instructions](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-testing-instructions-for-the-WooCommerce-Stripe-payment-gateway-5.8.0). I'll merge when the tests have succeeded 🎉

**Comment by asumaran** [comment]
@reykjalin the problem this PR fixes is very similar to (one of) the issues I've been trying to solve on PRBs (pcrvl0-e2-p2).

This approach will work just fine for simple products but I'm sure it won't work with variable products with different shipping types (normal/virtual) since we can't update the PRB to change the `requestShipping` option once the PRB is created. The only way to do that -I found so far- is to destroy the button and create it again with `requestShipping` the updated value. 

I just tested (develop 5d841445) with a product with two variations. One requires shipping and the other one doesn't. When the item that requires shipping is removed the payment sheet still asks for a Shipping Address even though the only added product is a virtual product which doesn't require shipping therefore a shipping address should not be required.

Screen Shot 2021-10-26 at 11 15 17 PM

**Pull request merged**